File: ts/src/stats/today.ts
// Copyright: Ankitects Pty Ltd and contributors
// License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html

import pb from "../backend/proto";

export interface TodayData {
    answerCount: number;
    answerMillis: number;
    correctCount: number;
    matureCorrect: number;
    matureCount: number;
    learnCount: number;
    reviewCount: number;
    relearnCount: number;
    earlyReviewCount: number;
}

const ReviewKind = pb.BackendProto.RevlogEntry.ReviewKind;

export function gatherData(data: pb.BackendProto.GraphsOut): TodayData {
    let answerCount = 0;
    let answerMillis = 0;
    let correctCount = 0;
    let matureCorrect = 0;
    let matureCount = 0;
    let learnCount = 0;
    let reviewCount = 0;
    let relearnCount = 0;
    let earlyReviewCount = 0;

    const startOfTodayMillis = (data.nextDayAtSecs - 86400) * 1000;

    for (const review of data.revlog as pb.BackendProto.RevlogEntry[]) {
        if (review.id < startOfTodayMillis) {
            continue;
        }

        // total
        answerCount += 1;
        answerMillis += review.takenMillis;

        // correct
        if (review.buttonChosen > 1) {
            correctCount += 1;
        }

        // mature
        if (review.lastInterval >= 21) {
            matureCount += 1;
            if (review.buttonChosen > 1) {
                matureCorrect += 1;
            }
        }

        // type counts
        switch (review.reviewKind) {
            case ReviewKind.LEARNING:
                learnCount += 1;
                break;
            case ReviewKind.REVIEW:
                reviewCount += 1;
                break;
            case ReviewKind.RELEARNING:
                relearnCount += 1;
                break;
            case ReviewKind.EARLY_REVIEW:
                earlyReviewCount += 1;
                break;
        }
    }

    return {
        answerCount,
        answerMillis,
        correctCount,
        matureCorrect,
        matureCount,
        learnCount,
        reviewCount,
        relearnCount,
        earlyReviewCount,
    };
}

File: rslib/src/stats/graphs/today.rs
// Copyright: Ankitects Pty Ltd and contributors
// License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html

use anki_proto::stats::graphs_response::Today;

use crate::revlog::RevlogReviewKind;
use crate::stats::graphs::GraphsContext;

impl GraphsContext {
    pub(super) fn today(&self) -> Today {
        let mut today = Today::default();
        let start_of_today_ms = self.next_day_start.adding_secs(-86_400).as_millis().0;
        for review in self.revlog.iter().rev() {
            if review.id.0 < start_of_today_ms {
                continue;
            }
            if review.review_kind == RevlogReviewKind::Manual
                || review.review_kind == RevlogReviewKind::Rescheduled
            {
                continue;
            }
            // total
            today.answer_count += 1;
            today.answer_millis += review.taken_millis;
            // correct
            if review.button_chosen > 1 {
                today.correct_count += 1;
            }
            // mature
            if review.last_interval >= 21 {
                today.mature_count += 1;
                if review.button_chosen > 1 {
                    today.mature_correct += 1;
                }
            }
            // type counts
            match review.review_kind {
                RevlogReviewKind::Learning => today.learn_count += 1,
                RevlogReviewKind::Review => today.review_count += 1,
                RevlogReviewKind::Relearning => today.relearn_count += 1,
                RevlogReviewKind::Filtered => today.early_review_count += 1,
                RevlogReviewKind::Manual | RevlogReviewKind::Rescheduled => unreachable!(),
            }
        }
        today
    }
}
